### PR TITLE
Components ES6 rewrite with improvements

### PR DESCRIPTION
**Pull request by Swiftb0y**
Reopen #4171 with Be-ing#56 on top. 

In the old PR, there is still a review from @Holzhaus which has not been addressed, though I plan to take care when I find the time.

### CODE REVIEW

**Comment by ywwg** [comment]
We should really be adding tests to our javascript corpus just like the rest of our codebase. Rather than commit this large PR all at once, I'd prefer to commit this a piece at a time, starting with just the base component class and accompanying tests that demonstrates its correctness, including communication with control objects and midi send/receive.  As a side-benefit, the test would also help serve as a way for controller writers to see how the library is supposed to be used.  

I had done some minor refactoring to make it easier to write controller tests here, but it was rejected because it was based on the old controller model: https://github.com/mixxxdj/mixxx/pull/3074.

**Comment by Swiftb0y** [comment]
IMO there are more pressing priorities for the controller right now. Before spending time adding the infrastructure to add tests, we should rather work on a newer mapping engine. Transitioning to QJSEngine was just the first part. Now we desperately need to get rid of the manual handler registration via XML and making ES6 Modules usable. From that point, adding tests is likely a matter of shipping a pre-made established library. However, before spending time on the controller engine, we should rather focus on Qt6 and thus the new QML UI... 

So what I'm trying to say is that there are more pressing issues right now and thus I'd like to keep the scope of this PR still as small as possible. This is mostly intended as a follow up PR to the QJSEngine transition which gave us ES7 support, so our APIs should leverage that. The functional changes are very minimal.

**Comment by ywwg** [comment]
Can you at least use `git mv` so it's clear what is the same as the old file and what's new?  As it is now it looks like 1100 lines of new code and it's not really feasible to review that much at once.  I can't just assume that it's a noop change without any sort of indication of testing or correctness.

**Comment by ywwg** [comment]
In general I'm not convinced by claims that we have "more important things to do than write tests."  Code is code and if we want Mixxx to remain stable, all code needs to be tested.  I spend a [lot](https://github.com/mixxxdj/mixxx/pull/4476) of [time](https://github.com/mixxxdj/mixxx/pull/4370) [fixing](https://github.com/mixxxdj/mixxx/pull/4093) [regressions](https://github.com/mixxxdj/mixxx/pull/3965), and that all subtracts from whatever new feature work I might want to do.  I am all for making ES6 classes usable, but if we're writing a library component that people are going to rely on, it needs some basic test coverage.

**Comment by Be-ing** [comment]
@ywwg Refusing to review code because it's above your arbitrary thresholds while demanding to drastically expand the scope of PRs is toxic. If you think hacking a testing framework into a legacy architecture is more important than moving to Qt6, well, have fun taking care of a dying application that still depends on Qt5 when the world is moving onto Qt7.

**Comment by Swiftb0y** [comment]
> Can you at least use git mv so it's clear what is the same as the old file and what's new?

We still have to ship the `0.0` version since many existing controller mappings depend on it.
The file here is a new version that got reiterated on (with some breaking changes since the new syntax is already a breaking change). using `git mv` does not make sense here. If you want to see the differences between the 0.0 and 0.1 version, I advise you to just diff the different files locally.

> In general I'm not convinced by claims that we have "more important things to do than write tests." Code is code and if we want Mixxx to remain stable, all code needs to be tested.

I generally agree, please don't get me wrong, I was not arguing against testing. I was saying that introducing the ability to even write tests is not worth the effort right now. Especially considering that most of the code is already "battletested" from the previous iteration.

**Comment by Swiftb0y** [comment]
@Be-ing, while I appreciate you are voicing your opinion here, please be more respectful. Owen does not have any obligation to review this PR, if he doesn't want to spend the time on reviewing either someone else will review this or this PR will start rotting like the rest of the 120 PRs.
@ywwg I'd still very much appreciate your input on this though. This is the opportunity for breaking changes so this is where we could start integrating features from your own JS library.

**Comment by Be-ing** [comment]
Of course, nobody has an obligation to review any particular PR. What I am taking issue with is this behavior of commenting on a PR to explicitly refuse to review it *while simultaneously demanding to drastically expand its scope*.

**Comment by Be-ing** [comment]
The new commits look fine to me. I leave it to @Holzhaus to follow up on his prior review.

**Comment by Be-ing** [comment]
By the way, thank you for the small, easy to review, self-contained commits with clear commit messages.

**Comment by ywwg** [comment]
@Swiftb0y , my worry is that it never seems to be a good time to write tests.  I honestly can't think of a better time to do so then when we're rewriting a library -- the workings of the code is in your mind and there is nothing depending on it yet (right?). I know there's a lot more exciting things to do but at some point we need to pay down this technical debt.


As for the library itself, the main reason I wanted to write mine was to separate the action of a button from the midi control it is connected to.  A button is two parts: the physical hardware with input and output, and the action that it takes and status that it gives.  With my little controller, buttons needed to have up to four different modes for a single piece of hardware.  I don't believe the current components library supports that level of indirection.  (An example from a non-custom hardware -- the Reloop MIXTOUR has three different shift buttons, and knobs and buttons take different actions depending which one is held).

**Comment by ywwg** [comment]
as for "increasing scope", I'm asking for the scope to be reduced -- Ideally we'd commit this file just starting with the base class and the button class and a few simple tests. then we add a few more subclasses.... and it would get pretty mechanical and move quickly.  There is the initial hurdle to write the first test, but that will pay off for *all* of our js files that need testing

**Comment by ywwg** [comment]
something like https://jestjs.io/

**Comment by Swiftb0y** [comment]
> my worry is that it never seems to be a good time to write tests.

I can understand that, but now is not only not the time but the arguments for doing it now have decreased weight:

>  I honestly can't think of a better time to do so then when we're rewriting a library -- the workings of the code is in your mind and there is nothing depending on it yet (right?).

Well this is not a full rewrite (the commit message is misleading). The behavioral aspects are the same (compare existing version with the new one). The main reason for this PR is that we modernize the API to use the new features language available with QJSEngine (replacing prototypical inheritance with class syntax, using arrow functions where applicable, using `for...of` instead of raw loops, etc). At the moment, there are 0 functional changes compared to the old library.

My point is that I don't want to force creators of new mappings to use old JS patterns because our library mandates it. It raises the barrier to entry for ComponentJS users. That is why I really want this PR to go into 2.4 (so its released together with QJSEngine). 

If we change existing code to add new features (rather than just adding a new component or fixing obvious bugs), then I'm also in favor of the gradual, test-centric approach, but making this a requirement for this PR does not make sense.

**Comment by Swiftb0y** [comment]
Thanks for your input in regards to what you'd like to see in a newer ComponentsJS version, we will consider them when doing an actual rewrite of the code.

**Comment by ywwg** [comment]
ok, and I appreciate the work. Hopefully there is someone already familiar with this code who can review it.

**Comment by uklotzde** [comment]
I am not a JavaScript developer and do not qualify for reviewing this PR.

If someone is willing to set up tests for the JavaScript parts please do so instead of demanding them by someone else. It is out of scope of this PR.

**Comment by ywwg** [comment]
Given the lack of anyone else being able to review this... I will give it a try.  When it's ready for review please mark it Ready For Review and I'll do my best.  thank you.

**Comment by Be-ing** [comment]
So I guess my review doesn't count?

**Comment by Holzhaus** [comment]
> So I guess my review doesn't count? Nor does @Holzhaus' review? @ywwg you have a [history](https://github.com/mixxxdj/mixxx/pull/2618#issuecomment-787528962) of belittling others' reviews. It's not okay. Writing "thank you" immediately after pretending like others' work doesn't exist comes across really badly.

9 days ago, @ywwg said that he hopes someone finds the time to review this. The only reaction to that was @uklotzde who said he doesn't feel qualified to review it. Nobody reviewed it since then, and there is still 0 formal reviews on this PR. 

So I agree with him that there currently is a "lack of people being able to review this", probably due to unfamiliarity with JS, time constraints, lack of motivation or other factors.

I appreciate that @ywwg wants to help to finish this PR and I don't feel belittled at all. His message sounded friendly to me. 

Besides: 10 days ago you attacked @ywwg and said "Refusing to review code [...] is toxic" (which implies that you agree that another review makes sense btw).  Now he agrees to review and you also attack him. I don't understand what you want.

**Comment by Be-ing** [comment]
I [did review this](https://github.com/mixxxdj/mixxx/pull/4550#issuecomment-986302662) yet @ywwg just posted "Given the lack of anyone else being able to review this". This is disrespectful.

**Comment by Swiftb0y** [comment]
Since the majority of the code is from you @Be-ing, merging this PR on the basis that you reviewed it is almost equivalent to merging your own PR.
Thank you @ywwg for offering to review this. I guess its reviewable in its current state. Any change can be added in separate PRs before the 2.4 release.

**Comment by Swiftb0y** [comment]
I guess this is a gray area to be clarified in our contribution guidelines.

**Comment by ywwg** [comment]
ok I'll take a look!

File: res/controllers/midi-components-0.1.js [hunk 1]
```diff
@@ -0,0 +1,1135 @@
...
+            return engine.getValue(this.group, this.outKey);
+        }
+        outSetValue(value) {
+            engine.setValue(this.group, this.outKey, value);
+        }
+        outToggle() {
+            this.outSetValue(!this.outGetValue());
+        }
+
+        connect() {
```
Comment (ywwg) [inline res/controllers/midi-components-0.1.js:107]: the block of functions above says they should not be overridden, but this one says it should.  Can we order the function definitions so override-ok functions are all on top, and the others below?  That way things most useful to the controller programmer are all in one place.

File: res/controllers/midi-components-0.1.js [hunk 2]
```diff
@@ -0,0 +1,1135 @@
...
+                this.connections[0] = engine.makeConnection(this.group, this.outKey, this.output.bind(this));
+            }
+        }
+        disconnect() {
+            for (const conn of this.connections) {
+                conn.disconnect();
+            }
+        }
+        trigger() {
+
```
Comment (ywwg) [inline res/controllers/midi-components-0.1.js:128]: nit: extra blank line

File: res/controllers/midi-components-0.1.js [hunk 3]
```diff
@@ -0,0 +1,1135 @@
...
+            for (const conn of this.connections) {
+                conn.trigger();
+            }
+        }
+        send(value) {
+            if (this.midi === undefined || this.midi[0] === undefined || this.midi[1] === undefined) {
```
Comment (ywwg) [inline res/controllers/midi-components-0.1.js:134]: this.midi isn't mentioned elsewhere, can we have a block of expected members at the top of the class, even if they are undefined?  That way programmers can know what values they can overwrite.

also, midi[0] and midi[1] are hard to understand without reading the implementation, can we make this something like midi.channel and midi.control?

File: res/controllers/midi-components-0.1.js [hunk 4]
```diff
@@ -0,0 +1,1135 @@
...
+                    midi.sendShortMsg(this.midi[0] + this.shiftOffset, this.midi[1], value);
+                } else if (this.shiftControl) {
+                    midi.sendShortMsg(this.midi[0], this.midi[1] + this.shiftOffset, value);
+                }
+            }
+        }
+    };
+    Component.prototype.max = 127;
+    Component.prototype.shiftOffset = 0;
+    Component.prototype.sendShifted = false;
```
Comment (ywwg) [inline res/controllers/midi-components-0.1.js:149]: I believe this means:  // if true, output values are sent not only to the designated midi channel and control, but also sent to associated midi controls. See below
// if shiftChannel is set, output values will be sent to the regular midi channel as well as the channel offset by this amount
// if shiftControl is set, output values will be sent to the regular midi control as well as the control offset by this amount.

Comment (ywwg) [inline res/controllers/midi-components-0.1.js:149]: I think this is where we should define this.midi so it's more clear how to use it, but perhaps this whole block should move to the top of the class.    Again, my thinking is that we want to have the parts the programmer is most likely to change at the top of the file, and the internal guts at the bottom.

**Review by ywwg — COMMENTED** [review]
This is the type of review I would like to write... is this helpful?  If this is a library we hope more people are going to use then I think we should work to make it friendlier.  I find it hard to follow right now and I think some basic rearrangement, comment documentation, and renaming of some things will make it better.

**Comment by ywwg** [comment]
Can you add a link to the wiki documentation at the top of the file? (https://github.com/mixxxdj/mixxx/wiki/Components-JS).  We'll have to bring it up to date but that will help people know where to start.

File: res/controllers/midi-components-0.1.js [hunk 5]
```diff
@@ -0,0 +1,1135 @@
+"use strict";
+/**
+ * Components JS library for Mixxx
+ * Documentation is on the Mixxx wiki at
+ * http://mixxx.org/wiki/doku.php/components_js
```
Comment (ywwg) [inline res/controllers/midi-components-0.1.js:5]: this link is broken, should be https://github.com/mixxxdj/mixxx/wiki/Components-JS

File: res/controllers/midi-components-0.1.js (hunk 2 above)
Comment (Swiftb0y) [reply res/controllers/midi-components-0.1.js:128]: I know, this actually was actually enforced by eslint, not sure why it trips up here.

File: res/controllers/midi-components-0.1.js (hunk 3 above)
Comment (Swiftb0y) [reply res/controllers/midi-components-0.1.js:134]: The purpose of the midi values are explained pretty clearly and in-depth in the wiki article. We could document them using jsdoc comments if you insist.

File: res/controllers/midi-components-0.1.js (hunk 4 above)
Comment (Swiftb0y) [reply res/controllers/midi-components-0.1.js:149]: You got it right. Sure I can leave comments, though are they really needed? `send()`  is so simple the code is pretty much self documenting imo.

Comment (Swiftb0y) [reply res/controllers/midi-components-0.1.js:149]: The issue is that there aren't really sensible default values for `this.midi`. Components might be used without outputting anything just to track some state or just for input handling (in the case for pots). 
We could make `this.midi` explicit by `Component.prototype.midi = [undefined, undefined]` (essentially just declaring the property)?

Reorderings in general should be fine, sure.

Comment (Be-ing) [reply res/controllers/midi-components-0.1.js:149]: `this.midi` can't be defined on the prototype. It is unique to each instance of Component.

Comment (Swiftb0y) [reply res/controllers/midi-components-0.1.js:149]: It can be defined there, but each instance needs to overwrite it regardless, so defining it there does not make a whole lot of sense as already pointed out.

Comment (Be-ing) [reply res/controllers/midi-components-0.1.js:149]: I once tried to move a property onto a prototype and have each instance override it, but that did not work because each instance was actually sharing the same data.

File: res/controllers/midi-components-0.1.js (hunk 2 above)
Comment (ywwg) [reply res/controllers/midi-components-0.1.js:128]: weird! ok

File: res/controllers/midi-components-0.1.js (hunk 3 above)
Comment (ywwg) [reply res/controllers/midi-components-0.1.js:134]: Now that I have a working link to the wiki I'll go through that.  Usually I prefer to keep documentation with the code so the docs stay up to date, but that can be a change we make later so we can leave it for now.

File: res/controllers/midi-components-0.1.js (hunk 4 above)
Comment (ywwg) [reply res/controllers/midi-components-0.1.js:149]: At least for myself, when I read a library API I'd prefer not to have to read the code to know how to use it.  My personal reference for best documentation is the google Abseil library ([example](https://github.com/abseil/abseil-cpp/blob/master/absl/time/civil_time.h)).  The comments are extensive and give lots of usage examples.  I'm not saying we have to do that much, but a few plain words in a comment can make it much easier for a programmer to understand what to do.  Remember that this particular library is meant to be used by many developers who are not familiar with Mixxx at all, so I believe we should do a little more work here to make it accessible to them.

Comment (ywwg) [reply res/controllers/midi-components-0.1.js:149]: hm ok.  This may just be a limitation of javascript then?

Comment (Swiftb0y) [reply res/controllers/midi-components-0.1.js:149]: Well, we already have extensive documentation in the form of the wiki page. Having the wiki page generated from docs and examples within the code would be great, but we don't have the facilities for that right now. Maybe jsdoc can help. I'll look into whether this is possible.

Comment (Swiftb0y) [reply res/controllers/midi-components-0.1.js:149]: > I once tried to move a property onto a prototype and have each instance override it, but that did not work because each instance was actually sharing the same data.

Thanks for the insight. I'll try to recreate this issue.

Comment (Be-ing) [reply res/controllers/midi-components-0.1.js:149]: I wouldn't say that's a limitation. That's just how prototypes work. If you don't want that, just set a normal property, not the prototype's property, which is what the code already does.

Comment (Be-ing) [reply res/controllers/midi-components-0.1.js:149]: Generating online documentation from inline comments would be nice but beyond the scope of this PR.

File: res/controllers/midi-components-0.1.js [hunk 6]
```diff
@@ -0,0 +1,1135 @@
...
+            // actual LED off.
+            if (this.colorKey !== undefined && outval !== this.off) {
+                this.outputColor(engine.getValue(this.group, this.colorKey));
+            } else {
+                this.send(outval);
+            }
+        }
+        outputColor(colorCode) {
+            // Sends the color from the colorCode to the controller. This
+            // method will not be called if no colorKey has been specified.
```
Comment (ywwg) [inline res/controllers/midi-components-0.1.js:316]: This comment is true now, but in general it's unsafe to make assumptions about how a function will be called if the prerequisites aren't enforced.  What if someone's controller script calls this function without a colorKey defined? If that's a problem, then we should check for colorKey != undefined and return if so.

File: res/controllers/midi-components-0.1.js [hunk 7]
```diff
@@ -0,0 +1,1135 @@
...
+                // Since outputColor has been called but no ColorMapper is
+                // available, we can assume that controller supports arbitrary
+                // RGB color output.
+                this.sendRGB(colorCodeToObject(colorCode));
+            }
+        }
+        sendRGB(_colorObject) {
+            // This method needs to be overridden in controller mappings,
+            // because the procedure is controller-dependent.
+            throw "sendRGB(colorObject) not implemented -- unable to send RGB colors";
```
Comment (ywwg) [inline res/controllers/midi-components-0.1.js:338]:
```suggestion
            throw "sendRGB(colorObject) not overridden -- unable to send RGB colors. Controller scripts must override this function with the implementation specific to that controller.";
```

File: res/controllers/midi-components-0.1.js [hunk 8]
```diff
@@ -0,0 +1,1135 @@
...
+        }
+        connect() {
+            super.connect();
+            if (this.group !== undefined && this.colorKey !== undefined) {
+                // TODO (Swiftb0y): replace with arrow function once https://bugreports.qt.io/browse/QTBUG-95677 got fixed
```
Comment (ywwg) [inline res/controllers/midi-components-0.1.js:343]: It's better not to assign ownership of TODOs so anyone can feel welcome to address them

```suggestion
                // TODO (https://bugreports.qt.io/browse/QTBUG-95677): replace with arrow function once this QT bug is fixed.
```

File: res/controllers/midi-components-0.1.js [hunk 9]
```diff
@@ -0,0 +1,1135 @@
...
+                this.connections[1] = engine.makeConnection(this.group, this.colorKey, function(color) {
```
Comment (ywwg) [inline res/controllers/midi-components-0.1.js:344]: does this override connection number two, or is it meant to append? This seems like a dangerous magic number.

File: res/controllers/midi-components-0.1.js [hunk 10]
```diff
@@ -0,0 +1,1135 @@
...
+    };
+
+    const SamplerButton = class extends Button {
+        constructor(options) {
+            if (options.number === undefined) {
+                throw "No sampler number specified for SamplerButton constructor";
+            }
+            options.group = "[Sampler" + options.number + "]";
+            options.outKey = null; // hack to get Component constructor to call connect()
+            super(options);
```
Comment (ywwg) [inline res/controllers/midi-components-0.1.js:360]: super constructors are always meant to be called before the child class does its own setup. And you've rightly identified that this is a hack.  Instead, can we think of a way to get this functionality to work without the hack?  (why can't we call connect() ourselves? I don't understand the code well enough)

File: res/controllers/midi-components-0.1.js [hunk 11]
```diff
@@ -0,0 +1,1135 @@
...
+                    this.send(this.off);
+                } else {
+                    this.send(this.empty);
+                }
+            }
+        }
+        connect() {
+            this.connections[0] = engine.makeConnection(this.group, "track_loaded", this.output.bind(this));
+            if (this.playing !== undefined) {
+                this.connections[1] = engine.makeConnection(this.group, "play", this.output.bind(this));
```
Comment (ywwg) [inline res/controllers/midi-components-0.1.js:418]: more worrying magic numbers.  Can we do this.connections.push?

File: res/controllers/midi-components-0.1.js [hunk 12]
```diff
@@ -0,0 +1,1135 @@
...
+        constructor(options) {
+            super(options);
+            this.firstValueReceived = false;
+        }
+        input(channel, control, value, _status, _group) {
+            if (this.MSB !== undefined) {
+                value = (this.MSB << 7) + value;
+            }
+            let newValue = this.inValueScale(value);
+            if (this.invert) {
```
Comment (ywwg) [inline res/controllers/midi-components-0.1.js:445]: we probably need to do value clamping before doing this inversion to protect against bad input

File: res/controllers/midi-components-0.1.js [hunk 13]
```diff
@@ -0,0 +1,1135 @@
...
+                }
+            });
+        }
+    };
+
+    const Deck = class extends ComponentContainer {
+        constructor(deckNumbers) {
+            super();
+            if (deckNumbers !== undefined) {
+                if (Array.isArray(deckNumbers)) {
```
Comment (ywwg) [inline res/controllers/midi-components-0.1.js:633]: this type deduction is a little gross, do we have different scripts doing different things?  Could we fix that?

(I am getting it in my head that we might want to migrate controller to typescript eventually, and the more type dynamism we have the harder that would be)

File: res/controllers/midi-components-0.1.js [hunk 14]
```diff
@@ -0,0 +1,1135 @@
...
+                    component.outKey = "group_" + newGroup + "_enable";
+                }
+            });
+        }
+        toggle() {
+            // cycle through deckNumbers array
+            let index = this.deckNumbers.indexOf(parseInt(
+                script.channelRegEx.exec(this.currentDeck)[1]
+            ));
+            if (index === (this.deckNumbers.length - 1)) {
```
Comment (ywwg) [inline res/controllers/midi-components-0.1.js:675]: simpler:  index = (index + 1) % (this.deckNumbers.length - 1)

File: res/controllers/midi-components-0.1.js [hunk 15]
```diff
@@ -0,0 +1,1135 @@
...
+                index = 0;
+            } else {
+                index += 1;
+            }
+            this.setCurrentDeck("[Channel" + this.deckNumbers[index] + "]");
+        }
+    };
+
+    const EffectUnit = class extends ComponentContainer {
```
Comment (ywwg) [inline res/controllers/midi-components-0.1.js:684]: Note: all of this effect stuff is *very* complex and would be my first target for some sort of testing in the future.

File: res/controllers/midi-components-0.1.js [hunk 16]
```diff
@@ -0,0 +1,1135 @@
...
+                                            "_Effect" + effectMatch[2] + "]";
+                        }
+                    }
+                });
+            };
+
+            this.toggle = function() {
+                // cycle through unitNumbers array
+                let index = this.unitNumbers.indexOf(this.currentUnitNumber);
+                if (index === (this.unitNumbers.length - 1)) {
```
Comment (ywwg) [inline res/controllers/midi-components-0.1.js:778]: mod trick again

File: res/controllers/midi-components-0.1.js [hunk 17]
```diff
@@ -0,0 +1,1135 @@
...
+                    index = 0;
+                } else {
+                    index += 1;
+                }
+                this.setCurrentUnit(this.unitNumbers[index]);
+            };
+
+            if (unitNumbers !== undefined) {
+                if (Array.isArray(unitNumbers)) {
```
Comment (ywwg) [inline res/controllers/midi-components-0.1.js:787]: more type deduction. Any way we can avoid?

File: res/controllers/midi-components-0.1.js [hunk 18]
```diff
@@ -0,0 +1,1135 @@
...
+                    };
+                    super(options);
+                }
+                // NOTE: This function is only connected when not in focus choosing mode.
+                onFocusChange(value, _group, _control) {
+                    if (value === 0) {
+                        if (colors !== undefined) {
+                            this.color = colors.unfocused;
+                        }
+                        this.group = "[EffectRack1_EffectUnit" +
```
Comment (ywwg) [inline res/controllers/midi-components-0.1.js:927]: we see this building of the group name over and over again -- can we factor that out into a function effectGroupName() or some sort?

File: res/controllers/midi-components-0.1.js [hunk 19]
```diff
@@ -0,0 +1,1135 @@
...
+                                true);
+                            if (!showParameters) {
+                                if (!allowFocusWhenParametersHidden) {
+                                    engine.setValue(this.group, "show_parameters", 1);
+                                    // eu.onShowParametersChange will refocus the
+                                    // previously focused effect and show focus in skin
+                                }
+                                this.pressedWhenParametersHidden = true;
+                            }
+                        } else {
```
Comment (ywwg) [inline res/controllers/midi-components-0.1.js:1052]: the nesting gets very deep here.  Where possible, can you do 
```
if { 
  ... 
  return;  
 }
...
```

instead of:
```
if {
...
} else {
...
}
```

**Review by ywwg — CHANGES_REQUESTED** [review]
OK I've finished going through the file.  The effect section is particularly complex and it's hard to know what's going on, so I'd like to flag that for eventual cleanup and testing.  But mostly I think it shouldn't be too much work to make some noticeable improvements.

thanks again for working on this.

File: res/controllers/midi-components-0.1.js (hunk 19 above)
Comment (Swiftb0y) [reply res/controllers/midi-components-0.1.js:1052]: Sure, I'm also favoring early returns, this was just not my code to begin with, but I can change it now surely.

File: res/controllers/midi-components-0.1.js [hunk 20]
```diff
@@ -0,0 +1,1135 @@
+"use strict";
+/**
+ * Components JS library for Mixxx
+ * Documentation is on the Mixxx wiki at
+ * https://github.com/mixxxdj/mixxx/wiki/Components-JS
```
[the remark this answers is not in the document]
Comment (Swiftb0y) [reply res/controllers/midi-components-0.1.js:5]: Yep, We should move this into to the manual as well imo. API docs like this benefit from versioning support a lot.

File: res/controllers/midi-components-0.1.js [hunk 21]
```diff
@@ -0,0 +1,1135 @@
...
+ *
+ *
+ * This library depends on Lodash, which is copyright JS Foundation
+ * and other contributors and licensed under the MIT license. Refer to
+ * the lodash.mixxx.js file in this directory for details.
+**/
+
+(function(global) {
+    const Component = class {
+        constructor(options) {
```
[the remark this answers is not in the document]
Comment (Swiftb0y) [reply res/controllers/midi-components-0.1.js:32]: I'll look into. idk how well JSDoc supports dynamic composition like that.

File: res/controllers/midi-components-0.1.js (hunk 6 above)
Comment (Swiftb0y) [reply res/controllers/midi-components-0.1.js:316]: Well this is assumed here because iirc, this method is essentially private. No user should call this themselves because the colorCode is an opaque type. I don't want to introduce much checks because this method is much hotter than most other parts of a component.

File: res/controllers/midi-components-0.1.js (hunk 9 above)
Comment (Swiftb0y) [reply res/controllers/midi-components-0.1.js:344]: `this.connections` is more like an array with slots for storing connections. IMO its a questionable design choice but I'm not sure how else to implement. Maybe a Symbols could help.

File: res/controllers/midi-components-0.1.js (hunk 10 above)
Comment (Swiftb0y) [reply res/controllers/midi-components-0.1.js:360]: Neither do I. I spend a couple minutes looking into this and just calling `connect` does not do the trick. I think the comment is a bit misleading because it doesn't actually depend on `connect` being called in the constructor but some of the other stuff thats called between the check for `outKey === undefined` and `connect()`.

File: res/controllers/midi-components-0.1.js (hunk 6 above)
Comment (ywwg) [reply res/controllers/midi-components-0.1.js:316]: Maybe prepend the function name with _ to indicate that it's "private"? That seems to be the convention in js-land.

Comment (ywwg) [reply res/controllers/midi-components-0.1.js:316]: ok, no need to add the undefined check

File: res/controllers/midi-components-0.1.js (hunk 13 above)
Comment (Swiftb0y) [reply res/controllers/midi-components-0.1.js:633]: > migrate controller to typescript eventually.

I'm very opposed to this for a number of reasons, but that discussion is out of scope. 

> this type deduction is a little gross, do we have different scripts doing different things? Could we fix that?

In this case, its essentially convenience. You could do the same in most statically typed languages (rust `enum`, haskell `Either`). Instead of having two different parameters, both of which could be unset, set with the correct type or incorrect type as well as being mutually exclusive. We just make use of the fact that those parameters have different types and then switch on type which actually simplifies the code.

File: res/controllers/midi-components-0.1.js (hunk 14 above)
Comment (Swiftb0y) [reply res/controllers/midi-components-0.1.js:675]: usually, `%` is expensive, but this code is not hot, so sure.

File: res/controllers/midi-components-0.1.js (hunk 15 above)
Comment (Swiftb0y) [reply res/controllers/midi-components-0.1.js:684]: I fully agree, When I did my review for the first commit from Be, I didn't touch this for the same reason. So thank you for taking the time here.

File: res/controllers/midi-components-0.1.js (hunk 17 above)
Comment (Swiftb0y) [reply res/controllers/midi-components-0.1.js:787]: Same reasoning in favor of it as above

File: res/controllers/midi-components-0.1.js (hunk 6 above)
Comment (Swiftb0y) [reply res/controllers/midi-components-0.1.js:316]: Sure, actually JS has support for members, but its too new of an ES version so we'll have to wait for [QTBUG-97724](https://bugreports.qt.io/browse/QTBUG-97724)

File: res/controllers/midi-components-0.1.js (hunk 8 above)
Comment (Swiftb0y) [reply res/controllers/midi-components-0.1.js:343]: I was not aware that this syntax was addressing anyone responsible for fixing things. I thought it was rather a signature so if someone wants to address the issue, they know who to talk to. Though to be fair, `git blame` would work just as well for that.

File: res/controllers/midi-components-0.1.js (hunk 21 above)
[the remark this answers is not in the document]
Comment (Be-ing) [reply res/controllers/midi-components-0.1.js:32]: If you want rigidly defined anything, you're in the wrong programming language. :P

File: res/controllers/midi-components-0.1.js (hunk 13 above)
Comment (Be-ing) [reply res/controllers/midi-components-0.1.js:633]: This is another case of "if you want rigidly defined types, use a different programming language". If you try to bend JS into something rigidly typed, you'll end up with quite strange and unidiomatic JS.

File: res/controllers/midi-components-0.1.js (hunk 15 above)
Comment (Be-ing) [reply res/controllers/midi-components-0.1.js:684]: If separating and loading libraries wasn't so clunky, I'd pull this out to separate library.

File: res/controllers/midi-components-0.1.js (hunk 9 above)
Comment (Be-ing) [reply res/controllers/midi-components-0.1.js:344]: It is meant to override. If you want to change this implementation detail to use `Array.push`, I think you'd need to clear the Array first.

File: res/controllers/midi-components-0.1.js (hunk 15 above)
Comment (Swiftb0y) [reply res/controllers/midi-components-0.1.js:684]: `components-0.2` task. I'll leave a todo

File: res/controllers/midi-components-0.1.js (hunk 6 above)
Comment (ywwg) [reply res/controllers/midi-components-0.1.js:316]: yeah underscore is not enforced but at least it's a signal

File: res/controllers/midi-components-0.1.js (hunk 8 above)
Comment (ywwg) [reply res/controllers/midi-components-0.1.js:343]: Oh interesting!  This is a minor point so no big whoop.  (I've also found attaching specific bugs to TODOs makes them more likely to get fixed)

File: res/controllers/midi-components-0.1.js (hunk 9 above)
Comment (ywwg) [reply res/controllers/midi-components-0.1.js:344]: Don't we need to disconnect the original connection before we overwrite it? Otherwise we've got a dangling connection that won't get removed if the script is reloaded.

File: res/controllers/midi-components-0.1.js (hunk 10 above)
Comment (ywwg) [reply res/controllers/midi-components-0.1.js:360]: thanks for taking a look at it.  We can leave this for now.

File: res/controllers/midi-components-0.1.js (hunk 9 above)
Comment (Be-ing) [reply res/controllers/midi-components-0.1.js:344]: It is the caller's responsibility to call `disconnect` before `connect`.

Comment (Swiftb0y) [reply res/controllers/midi-components-0.1.js:344]: Mhmm. we might want to introduce an abstraction for the `connections` then.

Comment (Be-ing) [reply res/controllers/midi-components-0.1.js:344]: I'm not sure what you mean. `disconnect` and `connect` are the abstractions for the callee. The `connections` array is not intended to be directly manipulated from the outside (and JavaScript has no way to enforce that...).

Comment (Swiftb0y) [reply res/controllers/midi-components-0.1.js:344]: The problem is that we need some way to talk about a connectionHandler (returned from `engine.makeConnection`) while being able to iterate over all owned by a particular object. I think @ywwg is concerned that these are currently just addressed by their index. We might instead want to consider making `connections` a `Map` or an `Object` so we can address these handlers using something more readable.
Would just using a plain `Object` suffice?

File: res/controllers/midi-components-0.1.js (hunk 14 above)
Comment (Swiftb0y) [reply res/controllers/midi-components-0.1.js:675]: The - 1 doesn't make sense in this context anymore. The version I pushed should be correct.

File: res/controllers/midi-components-0.1.js (hunk 9 above)
Comment (ywwg) [reply res/controllers/midi-components-0.1.js:358]: I am ok with making this a TODO as well -- we should come up with some way of handling connections that doesn't rely so much on callers doing the right thing.

File: res/controllers/midi-components-0.1.js (hunk 18 above)
Comment (ywwg) [reply res/controllers/midi-components-0.1.js:931]: I think this is the last bit of low-hanging fruit we can do now.  Does that seem reasonable?

**Review by ywwg — COMMENTED** [review]
Thanks for discussing these issues with me.  I resolved most of the issues and there's just a couple more things we talked about that should be easy to do. It's good to have a more solid idea of where we need to take this library to make it better!

**Comment by christophehenry** [comment]
This looks abandonned. Is there a reason?

**Comment by Swiftb0y** [comment]
no time / interested from me atm. you're welcome to pick this up if you'd like.
